rename to credential_configuration_id (#162)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

        |   interacts    |                                                    |
        |--------------->|                                                    |
        |                |                                                    |

>         |                |  (4) Token Request (Pre-Authorized Code, tx_code)  |

Reviewer: Should this be in a separate PR?

Reviewer 2: I think this is a left-over, might be addressed by existing PRs

Author: this happened because the branch used in Paul's PR was not up to date with main. will make sure there are no conflicts when merging.

This specification defines the following parameters for the JSON-encoded Credential Offer object:

* `credential_issuer`: REQUIRED. The URL of the Credential Issuer, as defined in (#credential-issuer-identifier), from which the Wallet is requested to obtain one or more Credentials. The Wallet uses it to obtain the Credential Issuer's Metadata following the steps defined in (#credential-issuer-wellknown).

> * `credential_configurations_offered`: REQUIRED. An array of unique strings that each identify one of the keys in the name/value pairs stored in the `credential_configurations_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.

Reviewer 3: ```suggestion
* `credential_configurations_offered`: REQUIRED. An array of unique strings, each representing a key in the `credential_configurations_supported` property of the Credential Issuer metadata. The Wallet utilizes these string values to access the corresponding object that holds information about the offered Credential, as defined in (#credential-issuer-parameters). For instance, these string values can be used to retrieve the scope value for use in the Authorization Request.
```

> * `credential_configurations_offered`: REQUIRED. An array of unique strings that each identify one of the keys in the name/value pairs stored in the `credential_configurations_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.

Reviewer: ```suggestion
* `credential_configurations_offered`: REQUIRED. An array of unique strings that each identify one of the Credential Configuration Identifiers in the `credential_configurations_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.
```

Author: ```suggestion
* `credential_configurations_offered`: REQUIRED. An array of unique strings that each identify one of the Credential Configuration Identifiers in the `credential_configurations_supported` Credential Issuer metadata property. The Wallet uses these string values to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, these string values can be used to obtain `scope` value to be used in the Authorization Request.
```

The request parameter `authorization_details` defined in Section 2 of [@!RFC9396] MUST be used to convey the details about the Credentials the Wallet wants to obtain. This specification introduces a new authorization details type `openid_credential` and defines the following parameters to be used with this authorization details type:

* `type`: REQUIRED. String that determines the authorization details type. MUST be set to `openid_credential` for the purpose of this specification.

> * `credential_configuration_id`: REQUIRED. String specifying a unique identifier of the Credential being described in the `credential_configurations_supported` map in the Credential Issuer Metadata as defined in (#credential-issuer-parameters). The referenced object in the `credential_configurations_supported` map conveys the details, e.g. format, for the requested Credential issuance. This specification defines Credential Format specific Issuer Metadata in (#format_profiles).

Reviewer 3: ```suggestion
* `credential_configuration_id`: REQUIRED. A string that uniquely identifies the Credential defined in the `credential_configurations_supported` map within the Credential Issuer Metadata, as detailed in (#credential-issuer-parameters). The corresponding object in the `credential_configurations_supported` map provides specific information, such as format, for the issuance of the requested Credential. This specification defines Credential Format specific Issuer Metadata in (#format_profiles).
```

Author: please do the bikeshedding in PR https://github.com/openid/OpenID4VCI/pull/155. this PR is strictly meant to change the parameter names only.

> * `credential_configuration_id`: REQUIRED. String specifying a unique identifier of the Credential being described in the `credential_configurations_supported` map in the Credential Issuer Metadata as defined in (#credential-issuer-parameters). The referenced object in the `credential_configurations_supported` map conveys the details, e.g. format, for the requested Credential issuance. This specification defines Credential Format specific Issuer Metadata in (#format_profiles).

Reviewer: ```suggestion
* `credential_configuration_id`: REQUIRED. String specifying the Credential Configuration Identifier that of the Credential being described in the `credential_configurations_supported` Credential Issuer Metadata parameter as defined in (#credential-issuer-parameters). The referenced object in the `credential_configurations_supported` parameter conveys the details, e.g. format, for the requested Credential issuance. This specification defines Credential Format specific Issuer Metadata in (#format_profiles).
```

Author: ```suggestion
* `credential_configuration_id`: REQUIRED. String specifying the Credential Configuration Identifier that of the Credential being described in the `credential_configurations_supported` Credential Issuer Metadata parameter as defined in (#credential-issuer-parameters). The referenced object in the `credential_configurations_supported` parameter conveys the details, such as format, for the issuance of the requested Credential. This specification defines Credential Format specific Issuer Metadata in (#format_profiles).
```

In addition to a mechanism defined in (#credential-authz-request), Credential Issuers MAY support requesting authorization to issue a Credential using OAuth 2.0 scope parameter.

> When the Wallet does not know which scope value to use to request issuance of a certain Credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials_configurations_offered` parameter values to identify object(s) in the `credential_configurations_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.

Reviewer: ```suggestion
When the Wallet does not know which scope value to use to request issuance of a certain Credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the Credential Configuration Identifiers in the `credentials_configurations_offered` parameter to identify the Credential Configurations in the `credential_configurations_supported` Credential Issuer metadata parameter and use the `scope` parameter value from that object.
```

Author: please introduce a new term "credential configuration identifier" in PR #155. this PR is strictly meant to change the parameter names only.

> When the Wallet does not know which scope value to use to request issuance of a certain Credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials_configurations_offered` parameter values to identify object(s) in the `credential_configurations_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.

Reviewer: ```suggestion
When the Wallet does not know which scope value to use to request issuance of a certain Credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credential_configurations_offered` parameter values to identify object(s) in the `credential_configurations_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.
```

Reviewer: fixed typo

In addition to the response parameters defined in [@!RFC6749], the AS MAY return the following parameters:

* `c_nonce`: OPTIONAL. String containing a nonce to be used when creating a proof of possession of the key proof (see (#credential_request)). When received, the Wallet MUST use this nonce value for its subsequent requests until the Credential Issuer provides a fresh nonce.
* `c_nonce_expires_in`: OPTIONAL. Number denoting the lifetime in seconds of the `c_nonce`.
* `authorization_details`: REQUIRED when `authorization_details` parameter is used to request issuance of a certain Credential type as defined in (#authorization-details). MUST NOT be used otherwise. An array of objects as defined in Section 7 of [@!RFC9396]. In addition to the parameters defined in (#authorization-details), this specification defines the following parameter to be used with authorization details type `openid_credential` in the Token Response:

>   * `credential_identifiers`: OPTIONAL. Array of strings, each uniquely identifying a Credential that can be issued using the Access Token returned in this response. Each of these Credentials corresponds to the same entry in the `credential_configurations_supported` Credential Issuer metadata but can contain different claim values or a different subset of claims within the claimset identified by that Credential type. This parameter can be used to simplify the Credential Request, as defined in (#credential_request), where the `credential_identifier` parameter replaces the format and any other Credential format-specific parameters in the Credential Request. When received, the Wallet MUST use these values together with an Access Token in subsequent Credential Requests.

Author: ```suggestion
  * `credential_identifiers`: OPTIONAL. Array of strings, each uniquely identifying a Credential that can be issued using the Access Token returned in this response. Each of these Credentials corresponds to the same entry in the `credential_configurations_supported` Credential Issuer metadata but can contain different claim values or a different subset of claims within the claimset identified by that Credential type. This parameter can be used to simplify the Credential Request, as defined in (#credential_request), where the `credential_identifier` parameter replaces the `format` parameter and any other Credential format-specific parameters in the Credential Request. When received, the Wallet MUST use these values together with an Access Token in subsequent Credential Requests.
```

>   * `credential_identifiers`: OPTIONAL. Array of strings that each uniquely identify a Credential instance that can be issued using Access Token returned in this response. Each Credential instance is a unique Credential described using the same entry in the `credential_configurations_supported` Credential Issuer metadata, but can contain different claim values or different subset of claims within the claimset identified by the Credential type. This parameter can also be used to simplify the Credential Request, since as defined in (#credential_request) `credential_identifier` parameter replaces `format` and any other Credential format specific parameters in the Credential Request. When received, the Wallet MUST use these values together with an Access Token in the subsequent Credential Request(s).

Reviewer: I still don't understand the difference and I think it is confusing since it conflates two concepts but this is probably just me. I hope the terminology PR will fix my lack of knowledge.

Author: @Sakurann need to fix few sentences here to clarify that there is no polymorphism. "the identifier refers to the instances of the credential configuration". drop `also`.

Author: ```suggestion
  * `credential_identifiers`: OPTIONAL. Array of strings that each uniquely identify a Credential that can be issued using Access Token returned in this response. Each Credential is described using the same object in the `credential_configurations_supported` Credential Issuer metadata, but can contain different claim values or different subset of claims within the claimset identified by that Credential type. This parameter can be used to simplify the Credential Request, since, as defined in (#credential_request), `credential_identifier` parameter replaces `format` and any other Credential format specific parameters in the Credential Request. When received, the Wallet MUST use these values together with an Access Token in the subsequent Credential Request(s).
```

@awoie @tlodderstedt please check if this is an improvement

Reviewer: ```suggestion
  * `credential_identifiers`: OPTIONAL. Array of strings, each uniquely identifying a Credential that can be issued using the Access Token returned in this response. Each of these Credentials corresponds to the same entry in the `credential_configurations_supported` Credential Issuer metadata but can contain different claim values or a different subset of claims within the claimset identified by that Credential type. This parameter can be used to simplify the Credential Request, as defined in (#credential_request), where the `credential_identifier` parameter replaces the format and any other Credential format-specific parameters in the Credential Request. When received, the Wallet MUST use these values together with an Access Token in subsequent Credential Requests.
```

A Client makes a Credential Request to the Credential Endpoint by sending the following parameters in the entity-body of an HTTP POST request using the `application/json` media type.

* `format`: REQUIRED when the `credential_identifier` was not returned from the Token Response. MUST NOT be used otherwise. String that determines the format of the Credential to be issued, which may determine the type and any other information related to the Credential to be issued. Credential Format Profiles consisting of the Credential format specific set of parameters are defined in (#format_profiles). When this parameter is used, `credential_identifier` parameter MUST NOT be present.

> * `proof`: OPTIONAL. Object containing the proof of possession of the cryptographic key material the issued Credential would be bound to.  The `proof` object is REQUIRED if the `proof_types` parameter is non-empty and present in the `credential_configurations_supported` map of the issuer metadata for the requested credential. The `proof` object MUST contain a following claim:

Reviewer 3: ```suggestion
* `proof`: OPTIONAL. Object containing the proof of possession of the cryptographic key material the issued Credential would be bound to.  The `proof` object is REQUIRED if the `proof_types` parameter is non-empty and present in the `credential_configurations_supported` map of the Issuer metadata for the requested Credential. The `proof` object MUST contain a following claim:
```

> * `proof`: OPTIONAL. Object containing the proof of possession of the cryptographic key material the issued Credential would be bound to.  The `proof` object is REQUIRED if the `proof_types` parameter is non-empty and present in the `credential_configurations_supported` map of the issuer metadata for the requested credential. The `proof` object MUST contain a following claim:

Reviewer: ```suggestion
* `proof`: OPTIONAL. Object containing the proof of possession of the cryptographic key material the issued Credential would be bound to.  The `proof` object is REQUIRED if the `proof_types` parameter is non-empty and present in the `credential_configurations_supported` Credential Issuer metadata for the requested credential. The `proof` object MUST contain a following claim:
```

Author: please do the bikeshedding in PR https://github.com/openid/OpenID4VCI/pull/155. this PR is strictly meant to change the parameter names only.

* `display`: OPTIONAL. An array of objects, where each object contains display properties of a Credential Issuer for a certain language. Below is a non-exhaustive list of valid parameters that MAY be included:
  * `name`: OPTIONAL. String value of a display name for the Credential Issuer.
  * `locale`: OPTIONAL. String value that identifies the language of this object represented as a language tag taken from values defined in BCP47 [@!RFC5646]. There MUST be only one object for each language identifier.

> * `credential_configurations_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is a unique identifier of the supported Credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific Credential and contains the following parameters defined by this specification:

Reviewer 3: ```suggestion
* `credential_configurations_supported`: REQUIRED. An object detailing the specifics of the Credential that the Credential Issuer is capable of issuing. This object comprises a list of name/value pairs, where each name uniquely identifies the supported Credential being described. This identifier is utilized in the Credential Offer, as outlined in (#credential_offer_parameters), to inform the Wallet about the offered Credential. The value is an object that contains metadata about a specific Credential and includes the following parameters defined by this specification:
```

Author: please do the bikeshedding in PR https://github.com/openid/OpenID4VCI/pull/155. this PR is strictly meant to change the parameter names only.

> * `credential_configurations_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is a unique identifier of the supported Credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific Credential and contains the following parameters defined by this specification:

Reviewer: ```suggestion
* `credential_configurations_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is a unique Credential Configuration Identifier of the supported Credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific Credential and contains the following parameters defined by this specification:
```

Author: please introduce a new term "credential configuration identifier" in PR #155. this PR is strictly meant to change the parameter names only.

Credential Issuers often want to know what Wallet they are issuing Credentials to and how private keys are managed for the following reasons:

> * The Credential Issuer MAY want to ensure that private keys are properly protected from exfiltration and replay to prevent an adversary from impersonating the legitimate Credential Holders by presenting their Credentials.

Reviewer 4: ```suggestion
* The Credential Issuer MAY want to ensure that private keys are properly protected from exfiltration and replay to prevent an adversary from impersonating the legitimate Credential Holder by presenting their Credentials.
```
Not sure, but I think we usually talk about a single Holder in these kind of statements.

If an adversary is able to get hold of a key proof defined in (#proof_types), the adversary could get a Credential issued that is bound to a key pair controlled by the victim.

> Note: For the attacker to be able to present to the Verifier a Credential bound to a replayed Key Proof, the attacker also needs to obtain the victim's private key. To limit this, servers are RECOMMENDED to check how the Wallet protects the private keys, using mechanisms such as Key Based Client Authentication defined in [@!I-D.ietf-oauth-attestation-based-client-auth].

Reviewer 3: ```suggestion
Note: For the attacker to be able to present to the Verifier a Credential bound to a replayed Key Proof, the attacker also needs to obtain the victim's private key. To limit this, servers are RECOMMENDED to check how the Wallet protects the private keys, using mechanisms such as Attestation-Based Client Authentication defined in [@!I-D.ietf-oauth-attestation-based-client-auth].
```

* Reference: (#token_request) of this specification

* Parameter Name: c_nonce
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token-response) of this specification

* Parameter Name: c_nonce_expires_in
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token-response) of this specification

## OAuth Dynamic Client Registration Metadata Registry

This specification registers the following client metadata name in the IANA "OAuth Dynamic Client Registration Metadata" registry [@!IANA.OAuth.Parameters] established by [@!RFC7591].

* Client Metadata Name: credential_offer_endpoint
* Client Metadata Description: Credential Offer Endpoint
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential_offer_endpoint) of this specification

> 

Reviewer: These should be removed since we have a separate PR that adds the IANA entries.

Author: this happened because the branch used in Paul's PR was not up to date with main. will make sure there are no conflicts when merging.

* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Provisional registration? No

* Type name: `application`
* Subtype name: `openid4vci-proof+cwt`
* Required parameters: n/a
* Optional parameters: n/a
* Encoding considerations: Binary CBOR, as specified in [@!RFC9052]
* Security considerations: See the Security Considerations in [@!RFC8392].
* Interoperability considerations: n/a
* Published specification: (#cwt-proof-type) of this specification
* Applications that use this media type: Applications that issue and store verifiable credentials
* Additional information:
  - Magic number(s): n/a
  - File extension(s): n/a
  - Macintosh file type code(s): n/a
* Person & email address to contact for further information: Torsten Lodderstedt, torsten@lodderstedt.net
* Intended usage: COMMON
* Restrictions on usage: none
* Author: Torsten Lodderstedt, torsten@lodderstedt.net
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Provisional registration? No

> 

Reviewer: Shouldn't these get removed since we have a separate PR for IANA registration?

Author: this happened because the branch used in Paul's PR was not up to date with main. will make sure there are no conflicts when merging.
